scripts/training/finetune-ldm-palette-adapter.py

```diff
@@ -0,0 +1,412 @@
+
+
+class AdapterLatentDiffusionConfig(BaseConfig):
+    """Finetunning configuration.
+
+    Contains the configs of the dataset, the latent diffusion model and the adapter.
+    """
+
+    dataset: DatasetConfig
+    ldm: LatentDiffusionConfig
+    adapter: AdapterConfig
+    test_ldm: TestPaletteDiffusionConfig
+
+
+class PaletteBatch(TypedDict):
+    """Structure of the data in the PaletteDataset."""
+
+    latent: Tensor
+    text_embedding: Tensor
+    palettes: dict[str, Tensor]
+
+
+class PaletteDataset(Dataset[PaletteBatch]):
+    """Dataset for the palette adapter.
+
+    Transforms the data from the Hugging Face dataset into PaletteBatches.
+    """
+
+    def __init__(self, trainer: "AdapterLatentDiffusionTrainer") -> None:
+        super().__init__()
+        self.trainer = trainer
+        self.dataset = self.load_huggingface_dataset()
+
+    @staticmethod
+    def download_image(
+        data: dict[str, list[Any]],  # "batched"
+        dl_manager: datasets.DownloadManager,
+    ) -> dict[str, list[str]]:
+        """Download the image from the url."""
+        urls = data["url"]
+        filenames: list[str] = dl_manager.download(urls)  # type: ignore
+        return {"image": filenames}
+
+    @staticmethod
+    def resize_image(
+        data: dict[str, list[Any]],  # "batched"
+        min_size: int = 512,
+        max_size: int = 576,
+    ) -> dict[str, list[Image.Image]]:
+        """Resize the shortest side of the image between `min_size` and `max_size`."""
+        return {
+            "image": [
+                resize_image(
+                    image=image,
+                    min_size=min_size,
+                    max_size=max_size,
+                )
+                for image in data["image"]
+            ],
+        }
+
+    @staticmethod
+    def encode_caption(
+        data: dict[str, list[Any]],  # "batched"
+        text_encoder: CLIPTextEncoderL,
+    ) -> dict[str, list[Tensor]]:
+        """Encode the caption with the encoder."""
+        return {
+            "text_embedding": [text_encoder(caption).squeeze(0) for caption in data["caption"]],
+        }
+
+    def load_huggingface_dataset(self) -> datasets.Dataset:
+        """Load the dataset from Hugging Face and apply some pre-processing."""
+        dataset_config = self.trainer.config.dataset
+        logger.info(
+            f"Loading dataset from {dataset_config.hf_repo}, "
+            f"revision {dataset_config.revision}, "
+            f"split {dataset_config.split}"
+        )
+        dataset = datasets.load_dataset(  # type: ignore
+            path=dataset_config.hf_repo,
+            revision=dataset_config.revision,
+            split=dataset_config.split,
+        )
+        dataset = dataset.select(list(range(100)))  # type: ignore # FIXME: temporary
+
+        # download images from urls
+        dl_manager = datasets.DownloadManager()
+        dataset = dataset.map(  # type: ignore
```

Comment:
How long does this take at cold start? (rough number)

Reply:
In my case, downloading the 24998 images of the [1aurent/unsplash-lite-palette](https://huggingface.co/datasets/1aurent/unsplash-lite-palette) dataset took about 20-30 minutes, though this obviously depends of your internet speed. Also, the images are a bunch of small files (about 2 to 5 MB each), so this doesn't help.
If not a cold start, this operation is instantaneous since the result of [`datasets.map`](https://huggingface.co/docs/datasets/main/en/package_reference/main_classes#datasets.Dataset.map) operations are cached.

Reply:
Good to know, thanks. Re: cache: true. Perhaps you want to expose the cache info settings (via the config) to choose a different path at some point.

Have you considered doing this offline once and for all and store a new dataset with image bytes bundled? Pros/cons?

Reply:
> Perhaps you want to expose the cache info settings (via the config) to choose a different path at some point.

The cache's location can be selected by modifying some environment variables, like [`HF_HOME`](https://huggingface.co/docs/huggingface_hub/main/en/package_reference/environment_variables#hfhome) or [`HF_DATASETS_CACHE`](https://huggingface.co/docs/datasets/cache#cache-directory).

> Have you considered doing this offline once and for all and store a new dataset with image bytes bundled? Pros/cons?

On advantage would be downloading the dataset in big chunks, thus potentially making the download faster.
If the images are bundled in the dataset, this would mean that images are loaded from the big `.arrow` files, and I don't know if this is faster than loading them from their respective `.png`/`.jpeg` files.
Though this doesn't matter much since after the `resize_images` mapping, the images are bundled inside `.arrow` files anyway. So this would only impact cold start.

Also the dataset's license doesn't allow redistributing the images, so I couldn't share it (publicly) on the huggingface hub.
> The Unsplash Dataset is made available for research purposes. [It cannot be used to redistribute the images contained within](https://github.com/unsplash/datasets/blob/master/TERMS.md).